### PR TITLE
Translation and edition of Java Binding User Guide

### PR DESCRIPTION
**Pull request by Maria-Nosyk**
## Overview
A translation of the Exonum Java Binding User Guide.

Link to the original document:
https://wiki.bf.local/display/~dmytro.tymofeev@bitfury.com/Exonum+Java+Binding+User+Guide

### CODE REVIEW

File: src/get-started/java-binding.md [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+or `Fork`, which is mutable and allows performing modifying operations. The
+database representation is provided by the framework: `Snapshot` can be
+requested at any time, while `Fork` - only when the transaction is executed. The
+lifetime of these objects is limited by the scope of the method to which they
+are transferred.
+
+Exonum stores arrays of bytes in the storage collections. Therefore, the user
+must implement serialization of values ​​stored in the collection. Java Binding
+provides *serializers* for standard and some commonly used types, see
+`com.exonum.binding.storage.serialization.StandardSerializers`.
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: see `com.exonum.binding.storage.serialization.StandardSerializers` for details.

File: src/get-started/java-binding.md [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+      var balanceById = ProofMapIndexProxy.newInstance(name, fork, StandardSerializers.hash(),
+          StandardSerializers.longs());
+      balanceById.put(id, newBalance);   
+    }
+    ```
+
+A set of named collections constitute a *service scheme*. For convenient access
+to service collections you can implement a factory of service collections.
+
+*The state of the service in the blockchain* is determined by the list of the
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: I'd put it as 'determined by the list of root hashes'

File: src/get-started/java-binding.md [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        return internalServerError()
+      ```
+
+#### Executable Transactions
+
+To correctly process a transaction, it must be transformed into an
+*executable transaction* (see `TransactionConverter` or
+`Service#convertTransaction` method) and transmitted to the framework using
+`Node#submitTransaction` method. The framework checks it, and if it is correct,
+broadcasts it to other nodes of the system. Other nodes, having received the
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: Many 'it' in this sentence, the sentence would be easier to understand if one of those was substituted by the full name of the object.

File: src/get-started/java-binding.md [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+`Transaction#execute` method describes the operations that are applied to the
+current storage state when the transaction is executed. Exonum transfers `Fork`
+as an argument - a view that allows performing modifying operations. A service
+schema object can be used to access data collections of this service.
+
+!!! note
+    `Transaction#info` will be deleted soon.
+
+For more details about transactions in Exonum - their properties and processing
+rules - see our [documentation][transactions]
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: see the corresponding section of our documentation/see the Transactions section.

File: src/get-started/java-binding.md [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+
+### Description of the External Service API
+
+The external service API serves for interaction of the service and the external
+systems. A set of operations is defined by each service and can include sending
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: The external service API is used for the interaction between the service and the external systems.

File: src/get-started/java-binding.md [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+transactions to the network, read requests for the blockchain data with the
+provision of the cryptographic proof, etc. Exonum provides an embedded web
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: read requests for blockchain data with the provision of cryptographic proof/of corresponding cryptographic proofs

File: src/get-started/java-binding.md [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+handler for balance requests in the "cryptocurrency" service will be available
+at `/cryptocurrency/balance/:walletId`.
+
+See [documentation][vertx.io] on the possibilities of `Vert.x` used as a web
+framework.
+
+### Dependencies Management
+
+To describe the dependencies of the service components (both system-specific
+ones, for example, Exonum time service, and external ones), Guice is used. Each
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: Guice is used to describe.../Exonum uses Guice to describe...
Generally putting the purpose first works well, but here the subject is quite far away from its verb.

File: src/get-started/java-binding.md [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+public class MySchemaIntegrationTest {
+  static {
+    LibraryLoader.load();
+  }
+
+  // Tests
+}
+```
+
+The plug-in for running the tests should be configured to transmit
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: for running tests

File: src/get-started/java-binding.md [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+    <configuration>
+        <argLine>
+            -Djava.library.path=${path-to-java-bindings-library}
+        </argLine>
+    </configuration>
+</plugin>
+```
+
+Each test can create `Snapshot`, `Fork`, using `MemoryDb`, and also apply `Fork`
+to its state:
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: Each test can create a `Snapshot`and/or `Fork`, using `MemoryDb`

Also, it's not very clear what 'its' refers to, so perhaps better to indicate the object to remove ambiguity.

File: src/get-started/java-binding.md [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+      @Test
+      public void testCreatingSnapshot() {
+        Snapshot snapshot = db.createSnapshot(cleaner);
+        // Use the Snapshot to test the schema
+      }
+
+      @Test
+      public void testMergingSomeChanges() {
+        Fork fork = db.createFork(cleaner);
+        // Make some changes to Fork
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: Missing dot at the end

File: src/get-started/java-binding.md [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+        // …
+
+        // Merge the changes made to the database Fork into the database
```
Comment (MartaDubno) [inline src/get-started/java-binding.md]: Missing dot at the end

File: src/get-started/java-binding.md [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+
+!!! note
+    `SparseListIndex` is not yet supported in Java. Let us know if it may be
+    useful for you!
+
+Collections work with a database view - either `Snapshot`, which is read-only;
+or `Fork`, which is mutable and allows performing modifying operations. The
+database representation is provided by the framework: `Snapshot` can be
+requested at any time, while `Fork` - only when the transaction is executed. The
+lifetime of these objects is limited by the scope of the method to which they
```
Comment (slowli) [inline src/get-started/java-binding.md]: Do I understand correctly that this lifetime restriction is true only for methods invoked by the core? That is, the `updateBalance` method below doesn't actually restrict the lifetime of its argument, right?

File: src/get-started/java-binding.md [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+Exonum stores arrays of bytes in the storage collections. Therefore, the user
+must implement serialization of values ​​stored in the collection. Java Binding
+provides *serializers* for standard and some commonly used types, see
+`com.exonum.binding.storage.serialization.StandardSerializers`.
+
+!!! note "Example of ProofMapIndex Creation"
+    ```java
+    void updateBalance(Fork fork) {
+      var name = "balanceById";
+      var balanceById = ProofMapIndexProxy.newInstance(name, fork, StandardSerializers.hash(),
```
Comment (slowli) [inline src/get-started/java-binding.md]: Nit: It could make sense to make this line shorter for the sake of mobile device friendliness.

File: src/get-started/java-binding.md [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+### Service Transactions Description
+
+Exonum transactions allow you to perform modifying atomic operations with the
+storage. Transactions are executed sequentially, in the order determined by the
+consensus of the nodes in the network.
+
+#### Messages
+
+Transactions are transmitted by the service client as an Exonum message. The
+transaction data in the message can be serialized using an arbitrary algorithm
```
Comment (slowli) [inline src/get-started/java-binding.md]: I'd rather say "the transaction payload"; if we consider transactions as a subtype of messages, some data in the transaction (e.g., `service_id`) must be serialized in a standard way.

File: src/get-started/java-binding.md [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+behavior and business logic. The interface implementations must define the
+transaction authentication rule (usually, the digital signature verification of
+the message) - `isValid` method; and the execution rule for the
+transaction - `execute` method.
+
+Ed25519 is a standard cryptographic system for digital signature of Exonum
+messages. We use `com.exonum.binding.crypto.CryptoFunctions#ed25519`
+method for obtaining the implementation of the function.
+
+The implementation of `Transaction#isValid` transaction authentication method
```
Comment (slowli) [inline src/get-started/java-binding.md]: Nit: `isValid` is not necessarily for authentication only; it can also verify internal constraints of the transaction (e.g., sender and receiver being separate entities in the transfer transaction in the cryptocurrency demo).

File: src/get-started/java-binding.md [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+        log("Internal error:" + e)
+        return internalServerError()
+      ```
+
+#### Executable Transactions
+
+To correctly process a transaction, it must be transformed into an
+*executable transaction* (see `TransactionConverter` or
+`Service#convertTransaction` method) and transmitted to the framework using
+`Node#submitTransaction` method. The framework checks it, and if it is correct,
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: "The framework _verifies_ it…"

_usually_, verifies the signature

File: src/get-started/java-binding.md [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+storage by external files or network resources in method implementations is
+not allowed.
+
+`Transaction#execute` method describes the operations that are applied to the
+current storage state when the transaction is executed. Exonum transfers `Fork`
+as an argument - a view that allows performing modifying operations. A service
+schema object can be used to access data collections of this service.
+
+!!! note
+    `Transaction#info` will be deleted soon.
```
Comment (slowli) [inline src/get-started/java-binding.md]: Hm, `info` has been removed from the core for a while; is there a particular reason to keep it in the binding?

File: src/get-started/java-binding.md [hunk 18]
```diff
@@ -0,0 +1,382 @@
...
+
+For more details about transactions in Exonum - their properties and processing
```
Comment (slowli) [inline src/get-started/java-binding.md]: I'd link to the article on transactions at the beginning of the "Service Transactions Description" section instead.

File: src/get-started/java-binding.md [hunk 19]
```diff
@@ -0,0 +1,382 @@
...
+rules - see our [documentation][transactions]
+
+### Description of the External Service API
+
+The external service API serves for interaction of the service and the external
+systems. A set of operations is defined by each service and can include sending
+transactions to the network, read requests for the blockchain data with the
+provision of the cryptographic proof, etc. Exonum provides an embedded web
+framework for describing the REST-interface of the service.
```
Comment (slowli) [inline src/get-started/java-binding.md]: "describing" probably isn't the right word here; there is no meta info on service endpoints thus far (which is what I'd think "describing the REST interface" means).

File: src/get-started/java-binding.md [hunk 20]
```diff
@@ -0,0 +1,382 @@
...
+
+To correctly process a transaction, it must be transformed into an
+*executable transaction* (see `TransactionConverter` or
+`Service#convertTransaction` method) and transmitted to the framework using
+`Node#submitTransaction` method. The framework checks it, and if it is correct,
+broadcasts it to other nodes of the system. Other nodes, having received the
+transaction message, convert it into an executable transaction, also using the
+service transaction converter.
+
+An executable transaction is a class object that implements
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: is an instance of a class implementing … ?

File: src/get-started/java-binding.md [hunk 21]
```diff
@@ -0,0 +1,382 @@
...
+`com.exonum.binding.messages.Transaction` interface and defines transaction
+behavior and business logic. The interface implementations must define the
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: defines its business logic ?

behaviour and business logic were used as synonyms, I think we can drop one here.

File: src/get-started/java-binding.md [hunk 22]
```diff
@@ -0,0 +1,382 @@
...
+transaction authentication rule (usually, the digital signature verification of
+the message) - `isValid` method; and the execution rule for the
+transaction - `execute` method.
+
+Ed25519 is a standard cryptographic system for digital signature of Exonum
+messages. We use `com.exonum.binding.crypto.CryptoFunctions#ed25519`
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: "Use <the method> to get its implementation" ?

We provide this method as part of our library.

File: src/get-started/java-binding.md [hunk 23]
```diff
@@ -0,0 +1,382 @@
...
+        node.submitTransaction(tx)
+        return ok(createTxResponse(hash))
+      catch IllegalArgumentException as e:
+        return badRequest("Invalid request: " + e)
+      catch VerificationError as e:
+        return badRequest("Invalid transaction: " + e)
+      catch Error as e:
+        log("Internal error:" + e)
+        return internalServerError()
+      ```
```
Comment (slowli) [inline src/get-started/java-binding.md]: Please un-indent the closing ticks by two spaces; the present formatting leads to incorrect rendering.

File: src/get-started/java-binding.md [hunk 24]
```diff
@@ -0,0 +1,382 @@
...
+    }
+    ```
+
+The full name of the module class must be submitted when the service launches.
+
+For more information on using Guice, see the [project wiki][Guice].
+
+### Testing
+
+#### Of Schema and Manipulations with the Storage
```
Comment (slowli) [inline src/get-started/java-binding.md]: Nit: I don't think a starting "of" is necessary for section names.

File: src/get-started/java-binding.md [hunk 25]
```diff
@@ -0,0 +1,382 @@
...
+        // Merge the changes made to the database Fork into the database
+        db.merge(fork);
+
+        // Create a Snapshot of the database state. Must include the changes
+        // made above.
+        Snapshot snapshot = db.createSnapshot(cleaner);
+        // Use the Snapshot to test the schema/transaction
+        // with a certain database state.
+        // …
+        }
```
Comment (slowli) [inline src/get-started/java-binding.md]: This and the following 2 lines should be un-indented for correct rendering.

File: src/get-started/java-binding.md [hunk 26]
```diff
@@ -0,0 +1,382 @@
...
+transaction - `execute` method.
+
+Ed25519 is a standard cryptographic system for digital signature of Exonum
+messages. We use `com.exonum.binding.crypto.CryptoFunctions#ed25519`
+method for obtaining the implementation of the function.
+
+The implementation of `Transaction#isValid` transaction authentication method
+must be a pure function, i.e. for the given transaction to return the same
+result on all nodes of the system. For this reason, access to the Exonum
+storage by external files or network resources in method implementations is
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: access to the Exonum storage, files or network resources …

File: src/get-started/java-binding.md [hunk 27]
```diff
@@ -0,0 +1,382 @@
...
+        // made above.
+        Snapshot snapshot = db.createSnapshot(cleaner);
+        // Use the Snapshot to test the schema/transaction
+        // with a certain database state.
+        // …
+        }
+      }
+      ```
+
+##### Of Transactions
```
Comment (slowli) [inline src/get-started/java-binding.md]: Why is this level-5 heading, rather than level-4 as surrounding ones?

File: src/get-started/java-binding.md [hunk 28]
```diff
@@ -0,0 +1,382 @@
...
+method for obtaining the implementation of the function.
+
+The implementation of `Transaction#isValid` transaction authentication method
+must be a pure function, i.e. for the given transaction to return the same
+result on all nodes of the system. For this reason, access to the Exonum
+storage by external files or network resources in method implementations is
+not allowed.
+
+`Transaction#execute` method describes the operations that are applied to the
+current storage state when the transaction is executed. Exonum transfers `Fork`
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: passes `Fork`?

File: src/get-started/java-binding.md [hunk 29]
```diff
@@ -0,0 +1,382 @@
...
+You can use [Vertx Web Client][vertx-web-client] as a client or another HTTP
+client.
+
+An example of API service tests can be found in
+`com.exonum.binding.cryptocurrency.ApiControllerTest`.
+
+### Known Limitations
+
+- Serialization is determined by a user, so Java services are not compatible
+  with JS light client (unless messages are updated).
```
Comment (slowli) [inline src/get-started/java-binding.md]: I believe "(unless messages are updated)" is an internal comment :upside_down_face:

File: src/get-started/java-binding.md [hunk 30]
```diff
@@ -0,0 +1,382 @@
...
+### Description of the External Service API
+
+The external service API serves for interaction of the service and the external
+systems. A set of operations is defined by each service and can include sending
+transactions to the network, read requests for the blockchain data with the
+provision of the cryptographic proof, etc. Exonum provides an embedded web
+framework for describing the REST-interface of the service.
+
+`com.exonum.binding.service.Service#createPublicApiHandlers` method is used to
+determine the handlers for HTTP requests. These handlers are available at the
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: to set the handlers/to initialize the handlers

File: src/get-started/java-binding.md [hunk 31]
```diff
@@ -0,0 +1,382 @@
...
+
+### Known Limitations
+
+- Serialization is determined by a user, so Java services are not compatible
+  with JS light client (unless messages are updated).
+- Core collections necessary to form a complete cryptographic proof for user
+  service data (collections and their elements) are available only in a "raw"
+  form - without deserialization of the content, which makes their use somewhat
+  difficult.
+- Not all system Rust services are available (EJB App configuration, Exonum
```
Comment (slowli) [inline src/get-started/java-binding.md]: I think you meant

> configuration, time and anchoring services will be integrated into EJB App

Also, I've got an impression that some services are integrated into the app already; is this the case?

File: src/get-started/java-binding.md [hunk 32]
```diff
@@ -0,0 +1,382 @@
...
+  time and anchoring services will be integrated).
+- Custom Rust services can be added to the application only by modifying and
+  rebuilding thereof.
+- The application supports only one Java service. However, there will be more!
```
Comment (slowli) [inline src/get-started/java-binding.md]: > However, there will be more!

Kind of awkward; I'd rephrase as "Support of multiple Java services is coming in the near future."

File: src/get-started/java-binding.md [hunk 33]
```diff
@@ -0,0 +1,382 @@
...
+        // Define the Service implementation.
+        bind(Service.class).to(CryptocurrencyService.class).in(Singleton.class);
+
+        // Define the TransactionConverter implementation.
+        bind(TransactionConverter.class).to(CryptocurrencyTransactionConverter.class);
+      }
+    }
+    ```
+
+The full name of the module class must be submitted when the service launches.
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: The fully-qualified name (or FQN) of the module class must be specified/passed during configuration of an Exonum App that will run the service.

?

File: src/get-started/java-binding.md [hunk 34]
```diff
@@ -0,0 +1,382 @@
...
+
+For more information on using Guice, see the [project wiki][Guice].
+
+### Testing
+
+#### Of Schema and Manipulations with the Storage
+
+To test the schema and manipulations with the storage, Exonum provides a
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: operations with the storage?

File: src/get-started/java-binding.md [hunk 35]
```diff
@@ -0,0 +1,382 @@
...
+public class MySchemaIntegrationTest {
+  static {
+    LibraryLoader.load();
+  }
+
+  // Tests
+}
+```
+
+The plug-in for running the tests should be configured to transmit
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: to pass `java.library.path` system property to JVM

File: src/get-started/java-binding.md [hunk 36]
```diff
@@ -0,0 +1,382 @@
...
+    <artifactId>maven-surefire-plugin</artifactId>
+    <configuration>
+        <argLine>
+            -Djava.library.path=${path-to-java-bindings-library}
+        </argLine>
+    </configuration>
+</plugin>
+```
+
+Each test can create `Snapshot`, `Fork`, using `MemoryDb`, and also apply `Fork`
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: apply changes made to `Fork` to the database state?

File: src/get-started/java-binding.md [hunk 37]
```diff
@@ -0,0 +1,382 @@
...
+#### Of Read Requests
+
+To test read requests for service data, you can use a fake that implements
+`Node` interface and uses `MemoryDb` to create `Snapshot`:
+`com.exonum.binding.service.NodeFake`. The `MemoryDb` contents can be filled in
+by executing `MemoryDb#merge(Fork)` operation as in the section above.
+
+#### Of API
+
+To test API implemented by Vertx tools, use the tools described in the
```
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: implemented with Vertx?

File: src/get-started/java-binding.md (hunk 29 above)
Comment (dmitry-timofeev) [inline src/get-started/java-binding.md]: I'd remove (unless messages are updated), because the core does not know when it will happen

File: src/get-started/java-binding.md [hunk 38]
```diff
@@ -0,0 +1,382 @@
+---
+title: Exonum Java Binding User Guide
+---
+# Exonum Java Binding User Guide
+
+Exonum Java Binding App is an application that includes the Exonum framework,
+system services and Java services runtime environment.
```
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply src/get-started/java-binding.md]: System services are the services that Exonum provides: configuration, anchroring, time oracle. They are not _currently_ bundled, but will be soon :)

File: src/get-started/java-binding.md [hunk 39]
```diff
@@ -0,0 +1,382 @@
...
+Download the archive with the [latest version of the application](https://github.com/exonum/exonum-java-binding/releases).
+
+**TBD:**  The following installation steps will be provided soon.
+
+## Java Services Development
+
+### Creating Project
+
+The easiest way to create a Java service project is to use the generic template
+generator. After installing Maven 3, run the command:
```
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply src/get-started/java-binding.md]: Exactly, will add in the installation steps

File: src/get-started/java-binding.md [hunk 40]
```diff
@@ -0,0 +1,382 @@
...
+
+The service abstraction serves to expand the framework and implement the
+business logic of the application. The service defines the schema of the stored
+data that constitute the service state; transaction processing rules that can
+make changes to the stored data; and the external API that allows to interact
+with the service from outside of the system. See more information on the
+software model of the service in the [corresponding section][Exonum-services].
+
+In Java, the abstraction of the service is represented by
+`com.exonum.binding.service.Service` interface. Implementations can use the
```
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply src/get-started/java-binding.md]: :+1: , We plan to publish generated Javadocs on our server, and update this guide with appropriate links.

File: src/get-started/java-binding.md (hunk 12 above)
Comment (dmitry-timofeev) [reply src/get-started/java-binding.md]: That is correct, but the only way to access an instance of a view _in production code_ is to have core call you (either Transaction#execute or the lambda passed to Node#withSnapshot).

Shall we additionally clarify that?

File: src/get-started/java-binding.md (hunk 15 above)
Comment (dmitry-timofeev) [reply src/get-started/java-binding.md]: Yes, but AFAIK we plan to move this functionality (message authentication) to the core framework, don't we? 

Also, there are currently three places where a user can check preconditions:
- a constructor
- `isValid`
- `execute`

Which is, IMO, a bad thing :) Therefore, I'd tell users to use it to authenticate the tx only until we move this method to core.

File: src/get-started/java-binding.md (hunk 17 above)
Comment (dmitry-timofeev) [reply src/get-started/java-binding.md]: It was renamed in the core, but in EJB we still use this name. By 'it will be deleted' I mean 'we consider getting rid of JSON altogether in the core framework'
